web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts

```diff
@@ -0,0 +1,240 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useEffect, useState } from 'react';
+import { useFileTransferContext } from 'shared/components/FileTransfer';
+
+import Tty from 'teleport/lib/term/tty';
+import { EventType } from 'teleport/lib/term/enums';
+import { Session } from 'teleport/services/session';
+import { DocumentSsh } from 'teleport/Console/stores';
+
+import { useConsoleContext } from '../consoleContextProvider';
+
+import { getHttpFileTransferHandlers } from './httpFileTransferHandlers';
+import useGetScpUrl from './useGetScpUrl';
+
+export type FileTransferRequest = {
+  sid: string;
+  requestID: string;
+  requester: string;
+  approvers: string[];
+  location: string;
+  filename?: string;
+  download: boolean;
+  isOwnRequest?: boolean;
+};
+
+export const useFileTransfer = (
+  tty: Tty,
+  session: Session,
+  currentDoc: DocumentSsh,
+  addMfaToScpUrls: boolean
+) => {
+  const { filesStore } = useFileTransferContext();
+  const ctx = useConsoleContext();
+  const user = ctx.getStoreUser();
+  const [fileTransferRequests, setFileTransferRequests] = useState<
+    FileTransferRequest[]
+  >([]);
+  const { getScpUrl, attempt: getMfaResponseAttempt } =
+    useGetScpUrl(addMfaToScpUrls);
+
+  useEffect(() => {
+    // the tty will be init outside of this hook, so we wait until
+    // it exists and then attach file transfer handlers to it
+    if (tty) {
+      tty.on(EventType.FILE_TRANSFER_REQUEST, updateFileTransferRequests);
+      tty.on(
+        EventType.FILE_TRANSFER_REQUEST_APPROVE,
+        handleFileTransferApproval
+      );
+      tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
+    }
+
+    return () => {
+      if (tty) {
+        tty.removeListener(
+          EventType.FILE_TRANSFER_REQUEST,
+          updateFileTransferRequests
+        );
+        tty.removeListener(
+          EventType.FILE_TRANSFER_REQUEST_APPROVE,
+          handleFileTransferApproval
+        );
+        tty.removeListener(
+          EventType.FILE_TRANSFER_REQUEST_DENY,
+          handleFileTransferDenied
+        );
+      }
+    };
+  }, [tty]);
```

Comment:
> which is using prevstate in the setState function which doesn't need any useCallback anyway.

Right, using `prevState` is the correct way to handle such case https://react.dev/reference/react/useEffect#updating-state-based-on-previous-state-from-an-effect.

But the outer function (for example `updateFileTransferRequests`) should still be memoized, although `setState` inside is stable. If you tried to access some other state in this function, you would get incorrect results because it would operate on the initial state value. Currently, you only update the state, so it works, but if we want to be 100% correct we should add memoize them.

Reply:
Sure sounds good. Will update today